### PR TITLE
ktfmt: init at 0.46

### PR DESCRIPTION
**Pull request by ghostbuster91**
## Description of changes

Add ktfmt package.
Homepage: https://github.com/facebook/ktfmt

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
It does not fit CONTRIBUTING.

Nixpkgs does not follow Semantic Commit messages. Reword the commit accordingly.

**Comment by ghostbuster91** [comment]
> It does not fit CONTRIBUTING.
> 
> Nixpkgs does not follow Semantic Commit messages. Reword the commit accordingly.

Sorry, my bad. Done.

File: pkgs/development/tools/ktfmt/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    repo = pname;
+    rev = "refs/tags/v${version}";
+    hash = "sha256-OIbJ+J5LX6SPv5tuAiY66v/edeM7nFPHj90GXV6zaxw=";
+  };
+
+  mvnHash = "sha256-pzMjkkdkbVqVxZPW2I0YWPl5/l6+SyNkhd6gkm9Uoyc=";
+
+  nativeBuildInputs = [ makeWrapper ];
+  installPhase = ''
+    mkdir -p $out/bin $out/share/ktfmt
```
Comment (eclairevoyant) [inline pkgs/development/tools/ktfmt/default.nix:18]:
```suggestion
    mkdir -p $out/bin
```

File: pkgs/development/tools/ktfmt/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    makeWrapper ${jre}/bin/java $out/bin/ktfmt \
+      --add-flags "-jar $out/share/ktfmt/ktfmt.jar"
+  '';
+
+  meta = with lib; {
+    description = "A program that reformats Kotlin source code to comply with the common community standard for Kotlin code conventions.";
+    homepage = "https://github.com/facebook/ktfmt";
+    license = licenses.apsl20;
+    mainProgram = "ktfmt";
+    maintainers = with maintainers; [ ghostbuster91 ];
```
Comment (eclairevoyant) [inline pkgs/development/tools/ktfmt/default.nix:30]: please add `meta.platforms`

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Comment by eclairevoyant** [comment]
also the commit message should be `ktfmt: init at 0.46`, as per https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
same with the PR title

File: pkgs/development/tools/ktfmt/default.nix (hunk 1 above)
Comment (ghostbuster91) [reply pkgs/development/tools/ktfmt/default.nix:18]: afaik `install` requires directories to exist and will return error if they are missing. Or are you suggesting that I should reuse the `bin` directory?

File: pkgs/development/tools/ktfmt/default.nix (hunk 2 above)
Comment (ghostbuster91) [reply pkgs/development/tools/ktfmt/default.nix:30]: done

File: pkgs/development/tools/ktfmt/default.nix (hunk 1 above)
Comment (eclairevoyant) [reply pkgs/development/tools/ktfmt/default.nix:18]: -D already creates the parent dirs

Comment (ghostbuster91) [reply pkgs/development/tools/ktfmt/default.nix]: done

File: pkgs/development/tools/ktfmt/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    owner = "facebook";
+    repo = pname;
+    rev = "refs/tags/v${version}";
+    hash = "sha256-OIbJ+J5LX6SPv5tuAiY66v/edeM7nFPHj90GXV6zaxw=";
+  };
+
+  mvnHash = "sha256-pzMjkkdkbVqVxZPW2I0YWPl5/l6+SyNkhd6gkm9Uoyc=";
+
+  nativeBuildInputs = [ makeWrapper ];
+  installPhase = ''
```
Comment (eclairevoyant) [inline pkgs/development/tools/ktfmt/default.nix:17]:
```suggestion
  installPhase = ''
  runHook preInstall
  
```

File: pkgs/development/tools/ktfmt/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    mkdir -p $out/bin
+    install -Dm644 core/target/ktfmt-*-jar-with-dependencies.jar $out/share/ktfmt/ktfmt.jar
+
+    makeWrapper ${jre}/bin/java $out/bin/ktfmt \
+      --add-flags "-jar $out/share/ktfmt/ktfmt.jar"
+  '';
```
Comment (eclairevoyant) [inline pkgs/development/tools/ktfmt/default.nix:23]:
```suggestion

  runHook postInstall
  '';
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
whoops, _almost_ missed this one last fix. see https://nix.dev/tutorials/learning-journey/packaging-existing-software#phases-and-hooks for context

**Comment by eclairevoyant** [comment]
also where is `jre_headless` coming from? I see only `jre` being passed in

**Comment by ghostbuster91** [comment]
> also where is `jre_headless` coming from? I see only `jre` being passed in, but `jre_headless` used in `meta.platforms`

right, I haven't tested that change :face_palm:

should be good now.

**Comment by ghostbuster91** [comment]
> whoops, _almost_ missed this one last fix. see https://nix.dev/tutorials/learning-journey/packaging-existing-software#phases-and-hooks for context

Thanks good to know. I was doing it based on ktlint that is also missing them. I will fix it in another PR.

**Review by eclairevoyant — APPROVED** [review]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ktfmt</li>
  </ul>
</details>

**Review by wegank — APPROVED** [review]

**Pull request merged**